feat: Add experimental support for peer-to-peer updates

gmaclennan opened the pull request:
This PR started as an attempt to add strict TypeScript type-checking to this code, but that work resulted in identifying several bugs and some issues with the existing architecture that created many opportunities for race conditions.

This is a really complicated piece of code, with some complicated logic for managing peers appearing and disappearing from a local network. It's important that we can know it works reliably, but also the code needs to be well documented and easy to maintain by the entire team. This PR attempts to address that.

In general I've found that writing this with Typescript and async / await has resulted in much faster development - the code here took about 25 hours to write (excluding thinking time away from the computer) and it should be easier and quicker to add features and fix bugs.

### Typescript

Everything is strictly typed now.

### State Management

We were using the module `rwlock` to manage async state management. The main async state that we need to maintain is that the service can start and stop many times (because the user can navigate to and from the sync screen, or close and re-open the app) and the start and stop functions can be async. The use of `rwlock` meant that the `starting` and `closing` states were hidden.

I created an abstraction, `AsyncService`, which manages start and stop of an async service according to tested rules. This means that we can test this complex code and simplify the code where the upgrade logic lies.

### async / await

Using callbacks with code that has so many async actions makes it hard to follow what is happening and catch race conditions. I have changed all the code to use async / await patterns, which I think makes the code easier to follow and maintain by avoiding nested callbacks and ensuring that the code flow is the same as the logic flow (top-to-bottom).

### State complexity

The shape of the `state` emitted by the UpgradeManager was complex, which resulted in a lot of code to decide what state to show to the user in the UX. The main things that the UX needs to know are:

1. Is the p2p upgrade service running (without error)?
2. Progress of any downloads
3. Progress of any uploads
4. Is there an update downloaded and available to install?

This PR simplifies the state to just capture that information, it is documented in `lib/types.d.js` as `UpgradeState`

### Code documentation

It's really important that this code is easy for others on the team to understand and maintain. It took me quite a while to parse the original code, so I have written many comments to explain why the code is written the way it is. I have also tried to be really clear about what are public methods and what methods are private (e.g. internal to the implementation and not something to be tested or used outside each class), with explanations of what each class and method does. I have used JSDoc format for comments, extended with Typescript annotations.

### Server

- No need for drain logic, since server.close() will stop accepting new connections, then wait for existing connections to close before closing.
- Include url property in installer schema returned by server (url is for downloading the APK)
- Fix start/stop logic in server to avoid race conditions
- Validate list installers response body matches schema
- Use fastify for better error handling and simpler API
- Simplified emitted state (only need a list of active uploads from server)

### Storage

- Simplify state logic and async to remove race conditions
- Read info about APKs from files themselves
- Cache installer info rather than read it each time
- Don't read files into memory for generating hash
- Use async methods for disk operations
- Turn into EventEmitter so can listen to available installers
- Async initialization to allow async cleanup of files (and reading info about current installers)
- createWriteStream() will only finish (e.g. emit "finish" event) when the installer is completely written to storage. This was tricky code, so I abstracted it as a util `startFinishStream()` and thoroughly tested it.

### Discovery

- Ensure onPeer event is removed when discovery is stopped, to avoid the event being added multiple times
- Add TTL & timeout to discovered installers from peers, so that installers from peers that go offline are removed from state
- Validate lists of installers from peers and protect against prototype pollution
- Simplified state that emits a list of available installers
- No logic about evaluating installers - will return all discovered installers no matter platform, version etc.
- Check port is available on each start (since it could become unavailable between stop() and start())
- createReadStream() for downloading (but abstracts how this is done) and ensures that an installer is removed from "available" if download fails (to avoid repeat attempts at download - will become available again if re-discovered by discovery.lookup())

### Evaluating potential upgrades

We had several places for comparing and evaluating upgrades. I tried to move all this logic into testable utility functions, with two separate tests:

1. Checking if a given installer is compatible with the current device
2. Getting the most recent of compatible installers

This should make it easier to add additional logic for selecting a suitable upgrade candidate.

### Next Steps

I really hope that these changes make this code more reliable and easier to maintain in the future.

There are a few tasks remaining (with estimated time):

- [x] Tests for upgrade-discovery (3h)
- [x] Tests for upgrade-server (2h)
- [x] Tests for upgrade-manager (4h)
- [x] Integration of new API with `server.js` (1h)
- [x] Update frontend code for new state shape (2h)
- [ ] Cleanup AsyncService code https://github.com/digidem/mapeo-mobile/pull/595#discussion_r635513490 (1h) - non-essential chore
- [x] Change all `_name` to `#name` https://github.com/digidem/mapeo-mobile/pull/595#discussion_r636315925 (1h) - non-essential chore
- [x] Evaluate how long to wait for initial download (1h)
- [x] Accidentally fixed beforeAfterStream tests for node@16, breaking node@12. Revert (1h)

When complete this code should close these issues: #589, #585, #576, #562, #559, #537

gmaclennan committed:
Add AppInfo native module for accessing sourceDir

gmaclennan committed:
Add share sheet for sharing Mapeo APK

gmaclennan committed:
Test auto-updating app from an APK

gmaclennan committed:
Test reading package signatures

hackergrrl committed:
Prep APK for sharing and start upgrade server.

hackergrrl committed:
get correct mapeo app version + show on settings

hackergrrl committed:
remove old comments

hackergrrl committed:
implement upgrade storage module

hackergrrl committed:
prevent 'filename' from being exposed on upgrades

hackergrrl committed:
add createReadStream API to stream upgrade data

hackergrrl committed:
chore: add eslint linting

hackergrrl committed:
test: move fake.apk to static/ subdir

hackergrrl committed:
test: formatting

hackergrrl committed:
fix: make 'arch' an array of strings

hackergrrl committed:
chore: add 'test' npm script

hackergrrl committed:
add initial UpgradeServer implementation

hackergrrl committed:
add new UpgradeServer implementation

hackergrrl committed:
test: add edge case tests for upgrade server

hackergrrl committed:
move the upgrade discovery key to its own file

hackergrrl committed:
have UpgradeServer accept a port to advertise on

hackergrrl committed:
make UpgradeStorage#getAvailableUpgrades sync

hackergrrl committed:
refactor: consolidate apk->upgradeoption logic

hackergrrl committed:
Revert "make UpgradeStorage#getAvailableUpgrades sync"

This reverts commit 2b4e3ac7dca59da5e8690e3c3038d8301ef102e9.

hackergrrl committed:
refactor: use a file for downloaded upgrade info

hackergrrl committed:
refactor: move LocalUpgradeInfo into a nodule

hackergrrl committed:
refactor: rename createApkWriteStream

hackergrrl committed:
add work-in-progress UpgradeDownloader component

hackergrrl committed:
chore: lint

hackergrrl committed:
add download progress tracking to state

hackergrrl committed:
implement the UpgradeDownloader checker component

hackergrrl committed:
chore: cleanup

hackergrrl committed:
implement the UpgradeManager

hackergrrl committed:
fix: ignore non-mapeo app keys

hackergrrl committed:
chore: add eslintrc for backend

hackergrrl committed:
don't access UpgradeDownloader's internals

hackergrrl committed:
auto-download upgrades that UpgradeManager finds

hackergrrl committed:
chore: lint

hackergrrl committed:
UpgradeDownload doesn't need version/arch/etc args

hackergrrl committed:
fix bugs in case where server waits to shut down

hackergrrl committed:
upgrade-server runs its own http server

hackergrrl committed:
expose 'startTime' context for upgrade searches

hackergrrl committed:
chore: upgrade deps

hackergrrl committed:
p2p-upgrades: pass in version to storage

hackergrrl committed:
wip: integrate upgrademanager into server

hackergrrl committed:
chore: add missing dep from develop

hackergrrl committed:
workaround for github:digidem/mapeo-mobile#521

hackergrrl committed:
wip: fixes

hackergrrl committed:
test: lint

hackergrrl committed:
fix: be able to listen for multiple ready events

hackergrrl committed:
add upload info to sharing state

hackergrrl committed:
DotIndicator: accept optional size parameter

hackergrrl committed:
wip: frontend implementation

hackergrrl committed:
move upgrade state constants to a common module

hackergrrl committed:
backend: upgrade simulator tool

hackergrrl committed:
implement upgrade state validation module

hackergrrl committed:
use a common storage path for downloaded APKs

hackergrrl committed:
expose filename property from upgrade-storage api

hackergrrl committed:
clone upgrade state context when modifying

hackergrrl committed:
add more thorough upgrade state management

hackergrrl committed:
fix: expose upgrade filename instead of hash

hackergrrl committed:
add more thorough state mgmt to upgrade-server

hackergrrl committed:
add more upgrade state mgmnt to upgrade-manager

hackergrrl committed:
wip: variety of frontend implementation upgrade changes

hackergrrl committed:
prevent older APKs from trying to be installed

hackergrrl committed:
simplify frontend state handling of upgrades

hackergrrl committed:
add another effect to track peer + upgrade state

hackergrrl committed:
test: fixes

hackergrrl committed:
test: files aren't left over after write error

hackergrrl committed:
feat: check hash upon upgrade download + cleanup

hackergrrl committed:
storage: use a temp dir for in-progress downloads

hackergrrl committed:
p2p-upgrades: make logging prefixes consistent

hackergrrl committed:
lint

hackergrrl committed:
add debug logging to upgrade backend

hackergrrl committed:
fix: write completed upgrade to root storage dir

gmaclennan committed:
Internationalization example

hackergrrl committed:
fix(upgrade): remove TTL; re-announce regularly

hackergrrl committed:
fix: readd flipper code (merge issue)

noffle committed:
Update src/backend/lib/local-upgrade-info.js

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

hackergrrl committed:
use strings for upgrade state constants

gmaclennan committed:
Internationalize text in update bar

gmaclennan committed:
Remove unused import

gmaclennan committed:
chore(linting): Combine duplicate imports

gmaclennan committed:
chore(linting): Remove unreachable code

gmaclennan committed:
Extract messages for translation

gmaclennan committed:
feat: Add minSdkVersion to AppInfo interface

gmaclennan committed:
chore: Add context to backend status errors

gmaclennan committed:
chore: Remove unused "flavor" option in createServer

gmaclennan committed:
chore: Don't try to support multiple "config" RPC messages

We wait for the config RPC message to start the server. This code would
try to restart the server with new config, but we don't have a use for
that and it makes the code a lot more complicated to manage, so I'm
removing it for now - it was not tested anyway and probably did not work

gmaclennan committed:
feat: Pass apkPath, minSdkVersion etc. to createServer()

apkPath: Path to currently running APK (file can be read from nodejs)
minSdkVersion: Min Android SDK version supported by currently running APK
version: Version name of running APK (semver if release, but not if testing)
buildNumber: Increments every time CI build happens - must always increase
bundleId: Different variants have different bundleIds
 - com.mapeo
 - com.mapeo.qa
 - com.mapeo.debug
 - com.mapeo.icca
Different bundle ids install as separate apps

hackergrrl committed:
java: add missing import

hackergrrl committed:
fix: prevent redownload of locally-stored upgrades

luandro committed:
chore: have UpgradeManager expose functions instead of events and ajust tests

hackergrrl committed:
do not copy the APK; use privateStorage instead

hackergrrl committed:
add optional param to upgrade simulator

gmaclennan committed:
chore: fix line-endings for gradlew.bat (should be CRLF)

gmaclennan committed:
Merge branch 'develop' into share-apk-p2p

* develop:
  feat: Setting to choose coordinate display format (#526)
  chore: Re-apply c267a99 with correct npm@6 version and lockfile v1
  Revert "feat: Updated translations for default configuration (vi, srn)"
  feat: Updated translations for default configuration (vi, srn)
  fix: Update Thai, Khmer & Vietnamese translations
  feat: Add Dutch translations
  feat: Add Sranan Tongo translations
  feat: Add French translations
  fix: Fix cutoff of text on OnePlus6T phone #502 (#511)
  fix: Write preset tags to observations (#514)
  chore: Fix Github CI builds by removing NDK r22 (#516)
  fix: Fix import config button crash (#512)
  feat: Updated translations (vi, es, po) (#498)

# Conflicts:
#	src/backend/package-lock.json
#	src/backend/package.json
#	src/frontend/screens/Settings/Settings.js

hackergrrl committed:
Revert "Test reading package signatures"

This reverts commit 28271b1a0166228bc05847208247bde55d3ccec7.

noffle committed:
Remove redundant keys from backend setup

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

noffle committed:
Merge pull request #566 from digidem/share-apk-p2p-gregor2

Add additional config available for UpgradeManager instance

hackergrrl committed:
remove unused error handling

gmaclennan committed:
Fix naming of apkVersion

gmaclennan committed:
Merge branch 'share-apk-p2p' into issue-536-2

* share-apk-p2p:
  Fix naming of apkVersion
  remove unused error handling
  Remove redundant keys from backend setup
  Revert "Test reading package signatures"
  feat: Pass apkPath, minSdkVersion etc. to createServer()
  chore: Don't try to support multiple "config" RPC messages
  chore: Remove unused "flavor" option in createServer
  chore: Add context to backend status errors
  feat: Add minSdkVersion to AppInfo interface

luandro committed:
Opt-in setting for p2p upgrades (#561)

* chore: cleanup old items in settings

* chore: add messages for experiments

* feat: Add Experiments page, route and state to SettingsContext

* feat: Only start p2p-upgrade if enabled on settings

* chore: Remove redundant code

* Add experimental flag to useEffect deps

Without this the useEffect will not run if this value changes

* fix: boolean type is lowercase

* Change shape of settings context

We plan to use this for other features in the future, so adding
`experiments` as a property on settingsContext.

* Set test IDs

Co-authored-by: Gregor MacLennan <gmaclennan@digital-democracy.org>

gmaclennan committed:
fix: Don't copy src/backend/test/fake.apk as part of CI Gradle build

gmaclennan committed:
Update backend build script to only copy whitelisted files

The backend build script uses noderify to bundle JS into a single file.
Previously the script would then blacklist unnecessary files to delete
them from the packaged app, but that is a maintenance burden.
This change whitelists the necessary files that must be included in the
APK and only includes those.

hackergrrl committed:
fix: use correct variable names

gmaclennan committed:
chore: Run backend tests on CI (#577)

* chore: Run backend tests on CI

* chore: Run backend tests in parallel with e2e tests

gmaclennan committed:
fix(AppInfo): Remove minSdkVersion

This property is only available in Android API 24 and later (Android v7)
We planned to use it to check compatibility incase our supported
minSdkVersion increased (it is currently API 20). Since this is not
available in phones we support, it is not useful for this purpose, and
using it causes a crash on phones running Android <v7

gmaclennan committed:
Revert "fix: use correct variable names"

This reverts commit 4526fba8e095ec1dc3e4abaa20b4310b884401e9.
Variable names were already corrected in 5709722a2fab65ab1fb330bd9566ca66bbfe5c3e

gmaclennan committed:
Merge branch 'share-apk-p2p' into issue-536-2

* share-apk-p2p:
  Revert "fix: use correct variable names"
  fix(AppInfo): Remove minSdkVersion
  chore: Run backend tests on CI (#577)
  fix: use correct variable names
  Update backend build script to only copy whitelisted files
  fix: Don't copy src/backend/test/fake.apk as part of CI Gradle build
  Opt-in setting for p2p upgrades (#561)

gmaclennan committed:
Add integration with server.js

gmaclennan committed:
Fix upgrade simulator for new upgrademanager pattern

gmaclennan committed:
Update settings store key for new settings shape

This will cause anyone who has used the "Coordinate format" setting to
loose their chosen format, but we have not released that yet. For the
next time we update the settings that are persisted, we need a way to
migrate old settings, because leaving this at @1 broke my phone, because
it had stored settings that did not have the experiments option.

gmaclennan committed:
Remove changes that should not be in this PR

gmaclennan committed:
Merge branch 'share-apk-p2p' into issue-536-2

* share-apk-p2p:
  Update settings store key for new settings shape

hackergrrl committed:
fix: upgrade semver sorting

hackergrrl committed:
fix: add upgrade options listener only once

hackergrrl committed:
fix: check for upgrades after current d/l finishes

hackergrrl committed:
fix: prioritize Downloading over ReadyToUpgrade

hackergrrl committed:
fix: reset upgrade state on download completion

hackergrrl committed:
fix: remove extra setState; fix param on a setState

hackergrrl committed:
fix: remove setState(IDLE) on error

hackergrrl committed:
test: fixes

These are some fixes re: recent changes to the upgrade state machines
that needed correction in the tests. These fixes are pretty hacky, but I
didn't feel it made sense to invest time into rewriting them right now.

hackergrrl committed:
fix: don't show search:IDLE as an error state

This is normal behaviour: the frontend will generally do a 'get-state'
request before search has initialized, revealing it as IDLE.

gmaclennan committed:
chore: Fix backend tests CI

hackergrrl committed:
fix: remove DOWNLOADED upgrade state

hackergrrl committed:
clean up UpgradeManager download logic

hackergrrl committed:
remove unused code

hackergrrl committed:
allow UpgradeCheck to re-check + choose latest version

gmaclennan committed:
fix: Throttle upload/download progress events to 100ms

noffle committed:
Merge pull request #580 from digidem/share-apk-p2p-redownload

Support downloading subsequent newer versions of the app

gmaclennan committed:
Merge commit '49b7464e2ab170939139f32a1026e037f4edf7b9' into issue-536-2

* commit '49b7464e2ab170939139f32a1026e037f4edf7b9':
  allow UpgradeCheck to re-check + choose latest version
  remove unused code
  clean up UpgradeManager download logic
  fix: remove DOWNLOADED upgrade state
  fix: don't show search:IDLE as an error state
  test: fixes
  fix: remove setState(IDLE) on error
  fix: remove extra setState; fix param on a setState
  fix: reset upgrade state on download completion
  fix: prioritize Downloading over ReadyToUpgrade
  fix: check for upgrades after current d/l finishes
  fix: add upgrade options listener only once
  fix: upgrade semver sorting
  chore: Fix backend tests CI

# Conflicts:
#	src/backend/lib/upgrade-manager.js

gmaclennan committed:
fix: Fix tests (maybe?)

gmaclennan committed:
Fix lint issues

gmaclennan committed:
fix: Throttle progress events to every 400ms

gmaclennan committed:
fix: Fix ApkInstaller for Android <7 (SDK version <24) (#586)

noffle committed:
Merge pull request #567 from digidem/issue-536-2

UpgradeManager expose functions

hackergrrl committed:
s/PROGRESS_THROTTLE/PROGRESS_THROTTLE_MS

noffle committed:
Merge pull request #588 from digidem/p2p-update/fix-progress-throttle

fix: Throttle upload/download progress events to 400ms

gmaclennan committed:
Use same node as nodejs-mobile

hackergrrl committed:
remove obsolete comment

hackergrrl committed:
Merge branch 'develop' into share-apk-p2p

gmaclennan committed:
upgrade-storage tests & fixes

gmaclennan committed:
rename state type

gmaclennan committed:
fix storage test

gmaclennan committed:
Remove hardcoded filepaths from expected APK info

hackergrrl commented:
I have yet to read it all through, but this looks like it's going to be a really valuable set of changes re: the reliability & maintainability of the backend codebase.

Honestly, it kinda makes me wanna just flip the switch and go full typescript on the MM backend >__>;

gmaclennan committed:
fix beforeAfterStream tests

gmaclennan committed:
UpgradeServer tests & corresponding fixes

gmaclennan committed:
Additional tests for trying to write invalid APKs (no uncaught error)

gmaclennan committed:
fix: Don't query a peer if a request to that peer is already in progress

hackergrrl commented:
@gmaclennan Can you rebase around this branch so that the diff only includes your changes & not also a bunch of merged changes from develop? This will make the diff a lot easier to read imo. Feel free to merge develop back into share-apk-p2p to make this happen!

gmaclennan committed:
Add UpgradeDiscovery test & fixes

(just one test for now, more coming later)

gmaclennan committed:
WIP: Refactor in Typscript

gmaclennan committed:
Fix upgrade candidate check for device.supportedAbis

gmaclennan commented:
@hackergrrl sure thing, now rebased.

gmaclennan committed:
delete upgrade-download.js (delete was lost in rebase)

gmaclennan committed:
Discovery test for peer going offline

gmaclennan committed:
oops, remove test.only()

gmaclennan committed:
Update Node version for CI tests (use same as nodejs-mobile)

gmaclennan committed:
Finish UpgradeDiscovery tests

gmaclennan committed:
Fix for backend build scripts

gmaclennan committed:
properly fix backend build scripts this time

gmaclennan committed:
Fix typo in comment

hackergrrl commented on src/backend/lib/async-service.js:
```diff
@@ -0,0 +1,199 @@
+        // Will still resolve once has started
+        return this.#started.promise;
+      case "error":
+        this.#started.reject(this.getState().error);
```
It seems like this code path could be hit multiple times once the service enters the "error" state: what happens if `reject()` gets called multiple times on the `#started` promise?

gmaclennan replied:
Good question! My understanding of promises is that once they are "fullfilled" (either rejected or resolved) then any further `resolve()` or `reject()` is a no-op. However the ECMA spec does not explicitly say anything like this. TL;DR `reject()` should be a no-op when called more than once. I can write a test to confirm this, or we could make our own deferred promise code that keeps state, e.g.

```js
function pDefer() {
  let fulfilled = false
  deferred.promise = new Promise((resolve, reject) {
    deferred.resolve = arg => {
      if (fulfilled) return
      fulfilled = true
      resolve(arg);
    }
    deferred.reject = arg => {
      if (fulfilled) return
      fulfilled = true
      reject(arg)
    }
  });
  return deferred;
```

However I don't think this is necessary.

gmaclennan replied:
Thanks for this comment. I added some tests for the error state behaviour and discovered a bug in the tests and a bug in the code. I think it is fixed now. This start() stop() seems simple but there are a lot of edge cases, so I'm happy we have this as an abstraction so we can test it in isolation.

hackergrrl replied:
Yes, I really liked the idea to separate out the async service bits so they can be reasoned about on their own.

hackergrrl commented on src/backend/lib/async-service.js:
```diff
@@ -0,0 +1,199 @@
+    this.#stopped = pDefer();
+    try {
+      this.setState({ value: "starting" });
+      await this._start.apply(this, args);
```
Shouldn't `#started` go back into a pending state before calling `_start()`? (Same Q for `stop` below)

gmaclennan replied:
We can only get here if `state.value === "stopped"` (because of switch above).
If `state.value === "stopped"` we guarantee that `#started.promise` is unfulfilled (e.g. pending) because the action `stopping` --> `stopped` (the only way to get to the `stopped` state) resets that promise.
The reset of `#started` needs to happen there, rather than here, because otherwise multiple calls to `start()` in the same tick would return different promises, and we would need to track and resolve all of them.
At least that all makes sense in my head! I tried to make sure the tests check all of this.

hackergrrl replied:
I see. Yeah, it's a bit difficult to track that mentally, but it makes sense!

I don't really like the separate management of the started/stopped promises; it seems fragile. I wonder if there's a way to encapsulate them so that the logic is more coherent. This might be a v2 thing though.

hackergrrl replied:
I haven't fully thought this through, but: what if `setState` was the only part that touched the started/stopped promises? That would centralize the logic & maybe make it easier to reason about.

gmaclennan replied:
Yes I agree, I was just thinking about this today - your comment highlighted a fragility here. As long as only this parent class manages state, everything is encapsulated and easier to reason about, but if a subclass were to change the state to "error" for example, that would break things. I will put on the todo to encapsulate everything in the parent class - I made a plan while driving to my parents'. I also agree that using promises for encapsulating this logic is harder to reason about, I think using an internal event emitter would be simpler (that is the approach I've used in the past, but this time I thought I'd try the deferred promise approach, which makes the code more concise, but at the expense of hiding state logic within promise states, so I don't like it on reflection).

gmaclennan replied:
Added to this PR's task list

hackergrrl commented on src/backend/lib/upgrade-discovery.js:
```diff
@@ -0,0 +1,211 @@
+    const installer = this.#availableInstallers.get(hash);
+    if (installer) {
+      const stream = got.stream(installer.installer.url);
+      stream.on("error", e => {
```
```suggestion
      stream.once("error", e => {
```

hackergrrl commented on src/backend/test/upgrade-discovery.test.js:
```diff
@@ -0,0 +1,268 @@
+    currentApkInfo
+  );
+  const discovery = new Discovery({
+    discoveryKey: "com.mapeo",
```
TODO: use bundleId to determine key

gmaclennan replied:
This is just in the test (hard-coded so that test instances discovery each other). In the code it used bundleId: https://github.com/digidem/mapeo-mobile/blob/p2p-update/typescript/src/backend/lib/upgrade-manager.js#L103
However, we should also test to make sure changing this means discovery does not happen.

gmaclennan replied:
Added a test for changing discovery keys (to ensure they don't discovery each other)

hackergrrl commented on .nvmrc:
```diff
@@ -1 +1 @@
-10.16
+12.16.3
```
You mentioned needing node > 12. Is this the new minimum?

gmaclennan replied:
The code uses a few JavaScript features that are only available in Node v12. I'm not sure how many are in Node >v12.0.0, but since the production code is running on Node v12.6.3 (due to the version of nodejs-mobile we use), I thought it best to target this specific version for development and testing.

hackergrrl reviewed (commented):
Initial comments. 13/46 files reviewed 💦

gmaclennan committed:
fix: Add tests from duplexify

gmaclennan committed:
Fix: once() on error

Co-authored-by: Kira Oakley <489362+hackergrrl@users.noreply.github.com>

gmaclennan committed:
Fix for error state of AsyncService

gmaclennan committed:
Add test to ensure different discovery keys do not find each other

hackergrrl commented on src/backend/lib/upgrade-discovery.js:
```diff
@@ -0,0 +1,211 @@
+
+// TTL for discovered peers, if they are not seen for more than this period then
+// they are considered offline and forgotten
+const TTL = 4000;
```
Nitpick:
```suggestion
const PEER_TTL_MS = 4000;
```

gmaclennan replied:
Good :)

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
-        this.upgradeOptions = [];
-      }
+// Throttle state events (in ms)
+const EMIT_THROTTLE = 400; // milliseconds
```
```suggestion
const EMIT_THROTTLE_MS = 400; // milliseconds
```

hackergrrl commented on src/backend/lib/upgrade-manager.js:
````diff
@@ -1,192 +1,275 @@
+ * }
+ * ```
+ *
+ * Will emit an "error" event if an upload or download fails. This is most
````
Maybe a note here about how transient vs fatal errors are handled? (It's not yet clear to me!)

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
+    availableUpgrade: undefined,
+  };
+  /** @type {Map<string, { progress: TransferProgress, stream: import('stream').Readable, installerInfo: InstallerExt }>} */
+  #downloads = new Map();
```
What is the difference between `#name` and `_name` use for member names?

gmaclennan replied:
Confession: I didn't know how to initialize [private instance fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields#private_instance_fields) from the constructor, so when I needed to initialize them there I used `_name`, which is just a visual indication (as opposed to `#name` which is enforced by the language). I need to switch these all over to:

```js
class ClassWithPrivateField {
  #privateField

  constructor() {
    this.#privateField = 42
  }
}
```

gmaclennan replied:
Added to this PR's task list

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
+   * @param {InstallerInt} options.currentApkInfo info about the currently installed APK
+   * @param {import('./types').DeviceInfo} options.deviceInfo Arch and SDK version of current device
+   */
+  constructor({ storageDir, currentApkInfo, deviceInfo }) {
```
Yay, glad `setApkInfo` was factored out.

gmaclennan replied:
I went backwards and forwards on this, and was initially passing in just the filepath to the current Apk, and reading the InstallerInfo within `UpgradeManager`, but that made the creation of the storage, server and discovery instances to be async, which made the code much more complicated, so now I'll change `server.js` to read that currentApkInfo from the file before the `UpgradeManager` instance is created.

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,268 @@
+  };
+}
+
+// If an APK does not have native code (e.g. no `lib` folder) then we assume it runs on these ABIs
```
Is this the default android bundling..?

gmaclennan replied:
Yes in my testing, if I create an empty Android app with no C++ code (eg only Java or Kotlin) then there is no "lib" folder. I assume that in that case the apk would run on any platform that Android runs on, but the main 4 (arm and x86, 32 and 64 bit) I think cover all likely platforms. In addition I don't see a future where we would ever release mapeo without native compiled code (it will either be node or rust I think)

gmaclennan committed:
fix: An installer with the same hash as current APK is not an upgrade

gmaclennan committed:
Fix: An installer < version of current APK is not an upgrade

gmaclennan committed:
fix: fake-apk-info should have same applicationId as valid APK fixtures

gmaclennan committed:
chore: Add initial upgrade manager integration test

gmaclennan committed:
fix: Don't download an upgrade candidate that we already have in storage

gmaclennan committed:
feat: Allow state emit throttle to be changed for testing

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
+   * @param {InstallerInt[]} storedInstallers
+   */
+  _onStoredInstallers(storedInstallers) {
+    const upgradeCandidate = getUpgradeCandidate({
```
I think that `getBestUpgradeCandidate` could be a more instructive name for this helper function?

gmaclennan replied:
Agreed! Changed.

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
+   * upgrade candidate
+   *
+   * @private
+   * @param {InstallerInt[]} storedInstallers
```
I like this concept of internal and external installer types -- good idea!

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
       }
-    });
-  }
+      // If there are any in progress downoads, cancel them
```
```suggestion
      // If there are any in progress downloads, cancel them
```

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -1,192 +1,275 @@
+    log("got request to start services");
+    const port = await getPort();
+    await Promise.all([this._discovery.start(port), this._server.start(port)]);
+    log("Started services");
```
Why does `_stop` need to call `setState` but `_start` doesn't?

gmaclennan replied:
Good catch! That is a bug, it should not be there in `_stop`, not sure how it sneaked in!

hackergrrl commented on src/backend/types/write-only-stream.d.ts:
```diff
@@ -0,0 +1,11 @@
+
+/// <reference types="node" />
+
+declare module "write-only-stream" {
```
What is this used for?

gmaclennan replied:
That is now unused. I used it in my first attempt at `beforeAfterStream.js`. Now removed.

hackergrrl commented on src/backend/types/gmaclennan_zip-fs.d.ts:
```diff
@@ -0,0 +1,15 @@
+// Project: https://github.com/gmaclennan/zip-fs
+// TypeScript Version: 4.2
+
+/// <reference types="node" />
```
What does this line do?

gmaclennan replied:
I don't fully understand these ["triple-slash directives"](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-path-) in Typescript, but I am copying this from other modules: it tells Typescript that the types reference types available in node.

hackergrrl reviewed (commented):
Another batch of code reviewed! 19/46 files so far :D

gmaclennan committed:
chore: rename var

Co-authored-by: Kira Oakley <489362+hackergrrl@users.noreply.github.com>

gmaclennan committed:
chore: rename var

gmaclennan committed:
chore: fix typo in comment

Co-authored-by: Kira Oakley <489362+hackergrrl@users.noreply.github.com>

gmaclennan committed:
fix: Remove incorrect setState in _stop()

gmaclennan committed:
chore: remove redundant dep

gmaclennan committed:
chore: rename function

gmaclennan committed:
chore: Comprehensive logging

gmaclennan committed:
chore: make all class fields private instance fields

I was using `_name` for instance fields initialized in the constructor.
This was not necessary. Changing all to `#name`.

gmaclennan committed:
chore: Additional APK fixtures for testing, and consistent naming scheme

gmaclennan committed:
chore: Setup UpgradeManager test scenarios

gmaclennan committed:
chore: integration tests for UpgradeManager

hackergrrl commented on src/backend/lib/upgrade-manager.js:
```diff
@@ -149,34 +168,77 @@ class UpgradeManager extends AsyncService {
+            0,
+            7
+          )} is downloading`
+        );
```
Such weird & imo unreadable auto-formatting! Is the line width configurable? Most of us probably use > 80 columns -- 120 might give improved readability.

gmaclennan replied:
Agreed! I like a lot of things about Prettier (it saves so much time for me when writing code), but I don't like how it formats long template literals. There is a long discussion about it: https://github.com/prettier/prettier/issues/3368
Let's tag the discussion about format line width for a discussion outside this PR, this is not something new added here.

hackergrrl commented on src/backend/lib/upgrade-storage.js:
```diff
@@ -1,271 +1,215 @@
+   * @returns {Promise<InstallerInt[]>}
+   */
+  async list() {
+    await this.started();
```
This is a nice pattern/trick to replace the locks. Very readable.

gmaclennan replied:
Glad it makes it readable.

gmaclennan committed:
chore: remove old tests

gmaclennan committed:
chore: Ensure tmp folders created in tests are removed afterwards

hackergrrl commented on src/backend/lib/upgrade-storage.js:
```diff
@@ -1,271 +1,215 @@
+    const forDeletion = new Set();
+
+    for (const installer of storedInstallerInts) {
+      if (!installer) continue;
```
Is this idea here to skip entries that got mapped to `{}`? If so, this won't work: `!!{} === true`.

gmaclennan replied:
`getInstallerInfo` either returns the `InstallerInt` type or throws. Line 71 catches this throw and returns undefined, so the entries of `storedInstallerInts` are either `void` or `InstallerInt` types. This skips `undefined` (`void`) entries, there are no entries mapped to `{}`.

hackergrrl replied:
I don't really understand this. I wrote this code to play with it:

```js
async function get (x) {
  if (x === -1) throw new Error('bad')
  else return true
}

const ps = [1,2,-1,3,-1].map(x => get(x).catch(() => {}))

Promise.all(ps)
  .then(res => {
    console.log('res', res)
  })
  .catch(err => {
    console.log('err', err)
  })
```
As you say, the `-1`s get mapped to `undefined`. But, if I change `{}` to `"foo"`, I get `"foo"` instead of `undefined`. Why does `{}` map to `undefined`??

gmaclennan replied:
This is JavaScript syntax being confusing. `() => {}` is a no-op function, the same as `function noop () {}`. The `{}` here is the empty body of a function, not an empty object. My apologies, this is a habit I've always used for writing a noop that may not be clear.

hackergrrl commented on src/backend/lib/upgrade-storage.js:
```diff
@@ -1,271 +1,215 @@
+    for (const installer of storedInstallerInts) {
+      if (!installer) continue;
+      // Mark any upgrade < current installed APK for deletion
+      if (installerCompare(installer, this.#currentApkInfo) === -1) {
```
Sanity check: this won't delete older versions of installers for different arches/platforms, right? e.g. We don't want to delete MapeoDesktop@3.0.0 just because we're running MapeoMobile@5.0.0.

gmaclennan replied:
Good catch. It raises an important question: what should the behaviour be for other arches/platforms? Keep the most recent of each platform?

hackergrrl replied:
Great question. I think this is worth filing an issue for to be thought about, but doesn't need to be answered yet / in this PR.

hackergrrl commented on src/backend/lib/upgrade-storage.js:
```diff
@@ -1,271 +1,215 @@
+          }
+          if (installer.hash !== expectedHash) {
+            // Hashes did not match, download was broken or incorrectly advertised
+            throw new Error("Invalid hash");
```
```suggestion
            throw new Error("Actual hash does not match expected hash");
```
The hash is [likely] a valid sha256 hash, but it's just not the value we expected it to be.

gmaclennan replied:
True! Thanks.

gmaclennan committed:
chore: Double-check downloaded hash in UpgradeManager tests

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,299 @@
+  // Important to validate semver before here, otherwise this will throw
+  const semverCompare = semver.compare(a.versionName, b.versionName);
+  if (semverCompare !== 0) return semverCompare;
+  if (a.versionCode < b.versionCode) return -1;
```
What is `versionCode`? Consider adding some comments on the `Installer{Ext,Int}` in `schema.js`.

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,299 @@
+  // Filter out the current APK
+  if (installer.hash === currentApkInfo.hash) return false;
+  // TODO: 64 bit devices can run 32 bit APKs
+  const isSupportedAbi = !!installer.arch.find(arch =>
```
Just a little simpler.
```suggestion
  const isSupportedAbi = installer.arch.some(arch =>
```

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,299 @@
+  if (installer.platform !== "android") return false;
+  if (installer.versionCode < currentApkInfo.versionCode) return false;
+  if (!semver.valid(installer.versionName)) return false;
+  if (installerCompare(installer, currentApkInfo) < 1) return false;
```
I notice we use this function a lot in the new codebase. For readability, consider some helper functions like `semver` offers, like `installerCompare.{lt,gt,eq}`. For me, I always need to pause and think for a few moments to reason about cmp function outputs.

gmaclennan replied:
That's a really good idea - I have had the same issue having to pause and think what `-1` vs `1` means.

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,299 @@
+ * 3. Version code must be greater or equal to currently installed app
+ * 4. Installer must be for platform 'android'
+ * 5. applicationId must match applicationId of currently installed app
+ * 6. versionName must be valid semver
```
What about `versionName` being newer than the current version?

gmaclennan replied:
Good point! @luandro can you add this to the comment in https://github.com/digidem/mapeo-mobile/pull/596

luandro replied:
:+1:

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,299 @@
+ *
+ * 1. Must support arch of device
+ * 2. Must support SDK of device
+ * 3. Version code must be greater or equal to currently installed app
```
Why equal? (I'm still not clear on what versionCode is, so that may also explain my uncertainty.)

gmaclennan replied:
You are correct, it should not be equal. Android will allow you to install an app with an equal version code, but if we allow that it will allow an inifinite loop of updates (imagine an APK with the same version name, e.g. `1.0.0` as the current app, and the same version code, but a different hash. It would be considered an update, but then in the other direction is also considered an update). We use build number for version code, so for Mapeo `versionCode` = build number. Build number is guaranteed to increment for every build of an APK we create on our CI.

gmaclennan replied:
Luandro is fixing this in https://github.com/digidem/mapeo-mobile/pull/596

luandro replied:
:+1:

hackergrrl commented on src/backend/lib/utils.js:
```diff
@@ -0,0 +1,299 @@
+ * @param {(wrappedStream: stream.Writable) => Promise<void>} options.finalize
+ * @returns {stream.Writable}
+ */
+function beforeAfterStream({ createStream, finalize }) {
```
This is some pretty tricky code to get 100% correct! I don't even feel totally qualified to evaluate its correctness. As far as I know, the write stream handling code I had in UpgradeStorage worked correctly. Why was this new [complex!] piece of code needed?

gmaclennan replied:
Yes this turned out to be more complex than I expected. I embarked upon it from a desire to create a simpler interface for UpgradeStorage, and didn't like the confusion of having to handle two callbacks - once when the write stream to tmp file finished, and once for when the file finally moves. In the end I think an interface like: `upgradeStorage.writeStream(readableStream, callback)` would have been simpler, even though it also creates a lack of clarity about where errors on `readableStream` should be handled. I lean heavily on tests here, and I copied the tests from `duplexify`. I feel confident that how we intend to use this helper, and the errors that we might get, are covered in the tests.

hackergrrl commented on src/backend/types/dns-discovery.d.ts:
```diff
@@ -0,0 +1,79 @@
+// Type definitions for dns-discovery 6.2.3
```
Did you generate these, or write them by hand? Curious what tool, if any, you used.

gmaclennan replied:
I write them by hand based on the docs, but often only typing the pieces that I use. Most modules already have open-source types in [Definitely Typed](http://definitelytyped.org) and can be installed by `npm i -D @types/dns-discovery`, but some less popular modules do not have community maintained types. It's confusing learning how to write these. You can just create a placeholder file with just the line: `declare module 'dns-discovery' {}` and Typescript will basically just ignore the module and not type-check your usage of it. I find writing these types helpful for sanity-checking how I am using the API of a module.

hackergrrl replied:
That seems like an interesting & useful exercise for building a stronger sense of understanding for how a module works.

hackergrrl reviewed (commented):
Implementation and types are reviewed. Just tests left. (38/54 files done.)

hackergrrl commented on src/backend/test/upgrade-manager.test.js:
```diff
@@ -0,0 +1,387 @@
+  sdkVersion: 21,
+};
+
+test.only("One device updates another", async t => {
```
^^^^

gmaclennan replied:
oops!

hackergrrl commented:
I notice that the sample APKs add another ~15mb to the git repo. Is there anything we can remove from these to make them smaller? I ask because any binary data we add to the repo (even if we modify or remove it in the future) is kept around forever.

hackergrrl commented:
I haven't had a chance to run this yet, but: how are fatal vs non-fatal errors handled on the frontend? I wasn't totally clear on the flow.

gmaclennan committed:
chore: Simplify abi check code

Co-authored-by: Kira Oakley <489362+hackergrrl@users.noreply.github.com>

gmaclennan committed:
chore: fix hash error message

Co-authored-by: Kira Oakley <489362+hackergrrl@users.noreply.github.com>

gmaclennan committed:
fix: Remove test.only()

gmaclennan commented:
> I notice that the sample APKs add another ~15mb to the git repo. Is there anything we can remove from these to make them smaller? I ask because any binary data we add to the repo (even if we modify or remove it in the future) is kept around forever.

I used the default empty project template in Android Studio. Just did a check and seems like turning on `minifyEnabled true` in `build.gradle` will reduce the size of each APK by 50%, so we could reduce this to ~8mb of APKs. Overall the git repo (including these APKs) is 75Mb right now. We can also use https://git-lfs.github.com although that adds a step for contributors

I can add a todo for re-generating these sample APKs minified to save space, then we can use https://rtyley.github.io/bfg-repo-cleaner/ to cleanup the larger ones.

gmaclennan commented:
> I haven't had a chance to run this yet, but: how are fatal vs non-fatal errors handled on the frontend? I wasn't totally clear on the flow.

Good point, will track that here: https://github.com/digidem/mapeo-mobile/issues/604

hackergrrl commented:
@gmaclennan Somehow (as an example), the [androidify base APK](https://github.com/node-on-mobile/androidify/blob/master/resources/base.apk) is only 230 kilobytes. Could be worth looking at for a few mins to see if there's something obvious we can do to reduce size.

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+  });
+  service.start();
+  service.start();
+  await new Promise(res => setTimeout(res, 0));
```
Why is this nextTick-like logic needed?

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+      await new Promise(res => setTimeout(res, 100));
+      startCount++;
+      // Throw the first time, but not on subsequent calls, but this should
+      // never be called after it has thrown once
```
If it shouldn't ever be called again after throwing, why have this guard around it? Wouldn't it be better to test it /wo any protections if the state should never happen?

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+  } catch (e) {
+    t.ok(e instanceof Error, "throws on first call");
+  }
+  t.equal(service.getState().value, "error", "in error state");
```
It would be nice to check that the state is "started" in the earlier tests also!

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+      started = false;
+    },
+  });
+  await service.start();
```
TODO: check the service's `state` after each operation?

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+  service.stop();
+  service.stop();
+  await new Promise(res => setTimeout(res, 0));
+  t.equal(stopCount, 0, "Service not yet started");
```
```suggestion
  t.equal(stopCount, 0, "Service not yet stopped");
```

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+      await new Promise(res => setTimeout(res, 20));
+      stopCount++;
+      // Throw the first time, but not on subsequent calls, but this should
+      // never be called after it has thrown once
```
Same comment here as above re: not needing an `if` guard.

hackergrrl commented on src/backend/test/async-service.test.js:
```diff
@@ -0,0 +1,286 @@
+  service.stop();
+  await nextTick();
+  await service.start();
+  t.true(started, "service is started");
```
Consider adding `getState` checks across these.

hackergrrl commented on src/backend/test/helpers.js:
```diff
@@ -0,0 +1,245 @@
+    const dstFilepath = path.join(tmpDir.path, path.basename(filepath));
+    await fsPromises.copyFile(filepath, dstFilepath);
+    expected.push({
+      // Just an empty object if readJson throws - sorry for ugly syntax
```
If `readJson` fails doesn't this mean something is VERY wrong with the requested test setup? Shouldn't this function throw in that case?

hackergrrl commented on src/backend/test/helpers.js:
```diff
@@ -0,0 +1,245 @@
+  readJson,
+  hashCmp,
+  setupStorage,
+  setupTmpStorageFolder,
```
This isn't used outside of this file -- remove from exports?

hackergrrl commented on src/backend/test/helpers.js:
```diff
@@ -0,0 +1,245 @@
+ * @param {number} [options.timeout=10000] After timeout, if `expectedValue` has not been emitted, the promise will timeout
+ * @param {string} [options.name=""]
+ * @param {import('type-fest').PartialDeep<T>} [options.always] Every event emitted whilst waiting for `expectedValue` should match this value
+ * @param {import('type-fest').PartialDeep<T>} [options.never] If any event emitted whilst waiting for `expectedValue`, promise will throw
```
```suggestion
 * @param {import('type-fest').PartialDeep<T>} [options.never] If any event emitted whilst waiting for `expectedValue` matches this value, the promise will throw
```

hackergrrl commented on src/backend/test/helpers.js:
```diff
@@ -0,0 +1,245 @@
+ * @param {string} eventName Name of event to await
+ * @param {T} expectedValue Promise will resolve once the emitted event partially matches this expected value
+ * @param {object} [options]
+ * @param {number} [options.timeout=10000] After timeout, if `expectedValue` has not been emitted, the promise will timeout
```
```suggestion
 * @param {number} [options.timeoutMs=10000] After timeout, if `expectedValue` has not been emitted, the promise will timeout
```

hackergrrl replied:
Suggesting using the suffix `Ms` for clarity that the units are milliseconds.

hackergrrl commented on src/backend/lib/types.d.ts:
```diff
@@ -0,0 +1,111 @@
+}
+// Info about the current device
+export interface DeviceInfo {
+  supportedAbis: InstallerExt["arch"];
```
What does this syntax do?

gmaclennan replied:
This selects the type of the property `arch` from the object/interface `InstallerExt`. Typescript does not support "dot" notation to access properties like you would on a Javascript object (e.g. `myObject.prop`, so you have to use `myObject["prop"]` when referring to props.

TL;DR this says that the `supportedAbis` prop has the same type as the `arch` prop of the `InstallerExt` type.

hackergrrl commented on src/backend/test/helpers.js:
```diff
@@ -0,0 +1,245 @@
+      // Sorry for ugliness! To save writing out apkInfo for availableUpgrades,
+      // we just pass in a filename of a JSON file describing this. This code
+      // reads the JSON and replaces the value `availableUpgrade`
+      const normalizedStep = { ...step };
+      for (const key of ["waitFor", "always", "never"]) {
+        // @ts-ignore
+        const value = step[key];
+        if (value) {
+          // @ts-ignore
+          normalizedStep[key] = { ...value };
+        }
+        if (value && typeof value.availableUpgrade === "string") {
+          // @ts-ignore
+          normalizedStep[
+            key
+          ].availableUpgrade = /** @type {InstallerInt} */ (await readJson(
+            path.join(validApksFolder, value.availableUpgrade)
+          ));
+        }
+      }
```
Please consider moving this into its own (pure) helper `normalizeStep()` function! It's pretty hairy and it'd be a good candidate for self-contained logic that explains what it's going and why.

hackergrrl replied:
Also, can you just use a module like `clone` here and append `availableUpgrade` to it afterwards? I cringe at the fragility of deep copy logic like this :)

gmaclennan replied:
Good idea, will add this to the todo.

hackergrrl commented on src/backend/lib/types.d.ts:
```diff
@@ -0,0 +1,111 @@
+
+interface ScenarioStepEvent {
+  /** Message passed to t.pass() */
+  message?: string;
```
Nitpick. You use VSCode right? Does it have refactor tools for these kinds of changes?
```suggestion
  passMessage?: string;
```

hackergrrl reviewed (commented):
More comments! 47/54 files reviewed.

gmaclennan commented:
> @gmaclennan Somehow (as an example), the [androidify base APK](https://github.com/node-on-mobile/androidify/blob/master/resources/base.apk?rgh-link-date=2021-05-25T16%3A50%3A57Z) is only 230 kilobytes. Could be worth looking at for a few mins to see if there's something obvious we can do to reduce size.

I had a look around today and found [this blog post](https://github.com/fractalwrench/ApkGolf/blob/master/blog/BLOG_POST.md) which helped me make a few small changes to the default Android app template (without any of the hacks) and got the APK size down to 11kb (so fairly significant reduction!). I've pushed it to a repo: https://github.com/digidem/minimal-android. It would be nice to create a script that generates APKs with different version names, version codes, minSdks and applicationIds for our tests. I can take a look at that later in the week.

FYI this is a useful tool for quickly inspecting an APK without Android Studio: https://apkanalyzer.netlify.app
I set that up on my own netlify account because the original is no longer working: https://github.com/bundlecop/apkanalyzer.io

gmaclennan committed:
WIP integrate new UpgradeManager API with server.js and front-end

Builds upon work in https://github.com/digidem/mapeo-mobile/pull/574

gmaclennan committed:
Add script to build APKs for testing p2p upgrades

gmaclennan committed:
Update Upgrade Simulator for new API

gmaclennan committed:
cleanup UpgradeBar code

gmaclennan committed:
Fix filepaths for share to switch to cache-path vs. files-path

hackergrrl commented on scripts/build-upgrade-test-apks.sh:
```diff
@@ -0,0 +1,35 @@
+# Build a QA Release APK with version name and version code
+build_apk 5.3.0 2000
+build_apk 5.4.0 2010
+build_apk 6.0.0 2020
```
Is the idea to manually edit these numbers when we do releases? This could be fragile!

gmaclennan replied:
This is just for testing. Version names are arbitrary, just needs to create 3 different versions for testing.

hackergrrl commented on src/backend/bin/upgrade-simulator.js:
```diff
@@ -1,43 +1,40 @@
+const { getInstallerInfo } = require("../lib/utils");
+
+const log = debug("p2p-upgrades:simulator");
+if (!process.env.DEBUG) debug.enable("p2p*");
```
Enable debug logging if NOT in debug mode?

hackergrrl commented on src/backend/bin/upgrade-simulator.js:
```diff
@@ -1,43 +1,40 @@
+const log = debug("p2p-upgrades:simulator");
+if (!process.env.DEBUG) debug.enable("p2p*");
+
+tmp.setGracefulCleanup();
```
💪🏻 😤

gmaclennan committed:
fix: Fix upload progress

gmaclennan committed:
Allow upgrade bar to expand in height if needed

gmaclennan committed:
Change button text to "Install" from "Install update" due to space con..
..straints

gmaclennan committed:
Don't show install button if there are active downloads / uploads

gmaclennan committed:
Add Mapeo version to wifi bar (above upgrade bar)

gmaclennan committed:
Create small button variant for use in upgrade bar

gmaclennan committed:
Add error icon for error state

gmaclennan committed:
Use normal weight text in upgrade bar

gmaclennan committed:
Cleanup unused imports

gmaclennan committed:
Turn off debug logging in production builds

gmaclennan commented:
@luandro, this should be ready for testing on a real device now. You will need to generate some Mapeo APKs with different versions in order to use the simulator, because the code now can't be fooled by advertising the APK as a different version, because it now checks the actual version of each APK.

If you are running the debug variant (which is what you do when you run `npm run android`), you can generate some different-versioned debug APKs to upgrade / downgrade with:

```sh
mkdir -p test-apks
cd android
ANDROID_VERSION_NAME=5.0.0 ANDROID_VERSION_CODE=1 ./gradlew assembleAppDebug
mv app/build/outputs/apk/app/debug/mapeo-app-debug.apk ../test-apks/mapeo-v5.0.0.1.apk
ANDROID_VERSION_NAME=6.0.0 ANDROID_VERSION_CODE=10 ./gradlew assembleAppDebug
mv app/build/outputs/apk/app/debug/mapeo-app-debug.apk ../test-apks/mapeo-v6.0.0.10.apk
```

That gives you two versions to test with: `v5.0.0` with a versionCode (build number) of `1` and `v6.0.0` with a versionCode `10`.

Now run Mapeo on your device specifying the exact version you want it to appear as:

```sh
ANDROID_VERSION_NAME=5.2.0 ANDROID_VERSION_CODE=5 npm run android
```

This will start Mapeo on your device as `v5.2.0` and versionCode `5`.

Now run the simulator with the older APK:

```sh
node ./src/backend/bin/upgrade-simulator.js test-apks/mapeo-v5.0.0.1.apk
```

You should see the device sharing its APK to the simulator, and in the simulator logs you should see the upgrade become available.

To test in the other direction (sharing a newer version from the simulator to the device):

```sh
node ./src/backend/bin/upgrade-simulator.js test-apks/mapeo-v6.0.0.10.apk
```

You should see the device download this version, then the UI will change to show an INSTALL button. This button should work to install the new version.

Let me know how (if!) this works for you, then we can roll out to the wider team for their testing. Because they won't be testing in the debug environment, I've created a script that will generate a few QA variant APKs with different version numbers for them to test:

```sh
./scripts/build-upgrade-test-apks.sh
```

We can run this when ready to generate a few APKs that the program team can install on different devices to test.

hackergrrl commented:
@gmaclennan Your notes here on how to build & set up the simulator and device together seem really useful & important! This could make great notes for a `README.md` or `HACKING.md` in the `backend/` dir.

hackergrrl commented on src/backend/lib/upgrade-server.js:
```diff
@@ -1,276 +1,195 @@
-  constructor(storage, port) {
+const EMIT_THROTTLE_MS = 400; // milliseconds
+
+// Only use logger in development when debugging
+// const serverLogger = process.env.DEBUG
+//   ? require("pino")({ prettyPrint: { singleLine: true } })
+//   : false;
```
Dead code?

gmaclennan replied:
Yes, but useful to keep for uncommenting during tests.

hackergrrl commented on src/backend/server.js:
```diff
@@ -173,30 +169,45 @@ function createServer({
-      rnBridge.channel.on("p2p-upgrade::get-state", () => manager.getState());
-      cb(null, manager);
+  function createUpgradeManager() {
+    // TODO: What is the effect of leaving these event handlers attached when
+    // the app is in the background? Ideally we would clean up event handlers
+    // each time the app goes to the background and re-attach them when back in
+    // foreground.
```
Consider filing this as an issue for better visibility & to make discussion easier.

hackergrrl commented on src/frontend/api.js:
```diff
@@ -98,6 +97,47 @@ export type ServerPeer = {
 
 type PeerHandler = (peerList: Array<ServerPeer>) => any;
 
+// These types are manually copied from `backend/lib/types.d.ts`. This is far
+// from ideal and is a workaround until we can convert the entire codebase to
+// Typescript - right now frontend uses Flow.
```
Consider filing an issue for this work :)

hackergrrl replied:
Also consider putting a comments in `backend/lib/types.d.ts` -- maybe in all-caps -- telling future devs to make sure any changes they make there are reflected in this file!

hackergrrl commented on src/frontend/api.js:
```diff
@@ -217,18 +257,21 @@ export function Api({
       const serverStartPromise = new Promise(resolve =>
         nodejs.channel.once("status", resolve)
-      ).then(() => {
+      ).then(async () => {
```
Is `async` required for this function? It doesn't `await` anywhere.

gmaclennan replied:
`deviceInfo` awaits on line 272

hackergrrl commented on src/frontend/screens/SyncModal/UpgradeBar.js:
```diff
@@ -0,0 +1,264 @@
+// @flow
```
Yay! This code refactored into its own file! 🌮

hackergrrl commented on src/backend/test/upgrade-discovery.test.js:
```diff
@@ -0,0 +1,299 @@
+const _ = require("lodash");
+const getStream = require("get-stream");
+
+const validApksFolder = path.join(__dirname, "./fixtures/valid-apks");
```
Doesn't `test/helpers.js` export this path already?

hackergrrl commented on src/backend/test/upgrade-storage.test.js:
```diff
@@ -0,0 +1,297 @@
+
+tmp.setGracefulCleanup();
+
+const validApksFolder = path.join(__dirname, "./fixtures/valid-apks");
```
Doesn't `test/helpers.js` export this path already?

hackergrrl commented on src/backend/test/upgrade-storage.test.js:
```diff
@@ -0,0 +1,297 @@
+tmp.setGracefulCleanup();
+
+const validApksFolder = path.join(__dirname, "./fixtures/valid-apks");
+const invalidApksFolder = path.join(__dirname, "./fixtures/invalid-apks");
```
Should `test/helpers.js` export this also?

hackergrrl commented on src/backend/test/utils/getInstallerInfo.test.js:
```diff
@@ -0,0 +1,41 @@
+const { promises: fs } = require("fs");
+
+test("getInstallerInfo: read valid APKs", async t => {
+  const validApksFolder = path.join(__dirname, "../fixtures/valid-apks");
```
Doesn't `test/helpers.js` export this path already?

hackergrrl commented on src/backend/test/utils/getInstallerInfo.test.js:
```diff
@@ -0,0 +1,41 @@
+
+test("getInstallerInfo: invalid APKs", async t => {
+  const invalidApksFolder = path.join(__dirname, "../fixtures/invalid-apks");
+  const apkFilepaths = (await fs.readdir(invalidApksFolder))
+    .filter(file => path.extname(file) === ".apk")
+    .map(file => path.join(invalidApksFolder, file));
```
If you're using this pattern a lot, it could stand to be a parsing job that helpers.js does on init + exports.

hackergrrl commented on src/backend/test/utils/getInstallerInfo.test.js:
```diff
@@ -0,0 +1,41 @@
+
+async function readJson(filepath) {
+  return JSON.parse(await fs.readFile(filepath));
+}
```
`test/helpers.js` exports this!

hackergrrl commented on src/backend/test/upgrade-storage.test.js:
```diff
@@ -0,0 +1,297 @@
+  );
+
+  // Check both list() and "installers" event
+  const [installersFromList, [installersFromEvent]] = await Promise.all([
+    storage.list(),
+    once(storage, "installers"),
+  ]);
+
+  t.deepEqual(
+    installersFromList,
+    [fakeApkInfo.v0_0_1],
+    "Starts with only current apk"
+  );
+  t.deepEqual(
+    installersFromEvent,
+    [fakeApkInfo.v0_0_1],
+    "Starts with only current apk"
+  );
```
I really like how easy + readable it is to express this kind of logic!

hackergrrl commented on src/backend/test/upgrade-storage.test.js:
```diff
@@ -0,0 +1,297 @@
+    await pipeline(fs.createReadStream("non-existent-file.apk"), ws);
+  } catch (e) {
+    t.ok(e instanceof Error, "throws an error");
+  }
```
Consider using e.g. `t.plan()` to set how many positive signals you expect. Otherwise, if the pipeline **doesn't** throw, this test will still pass.

hackergrrl commented on src/backend/test/upgrade-storage.test.js:
```diff
@@ -0,0 +1,297 @@
+    await pipeline(fs.createReadStream(testApkFilepath), ws);
+  } catch (e) {
+    t.ok(e instanceof Error, "writeStream throws");
+  }
```
Same feedback as the previous test.

hackergrrl commented on src/backend/test/upgrade-storage.test.js:
```diff
@@ -0,0 +1,297 @@
+    try {
+      await pipeline(fs.createReadStream(filepath), ws);
+    } catch (e) {
+      t.ok(e instanceof Error, "throws an error");
```
Consider using e.g. t.plan() to set how many positive signals you expect. Otherwise, if the pipeline doesn't throw, this test will still pass.

hackergrrl commented on src/backend/test/upgrade-server.test.js:
```diff
@@ -0,0 +1,204 @@
+const fsPromises = fs.promises;
+const got = require("got").default;
+
+tmp.setGracefulCleanup();
```
Can this be done just *once* in `helpers.js`? It should affect all test files (that include the helpers.js file, which I think is all of them).

gmaclennan replied:
Not sure, risky given how different test runners might work.

hackergrrl commented on src/backend/test/upgrade-server.test.js:
```diff
@@ -0,0 +1,204 @@
+  };
+  const { address, port } = /** @type {import('net').AddressInfo} */ (start
+    ? app.server.address()
+    : { address: "localhost", port: 80 });
```
Is this code path used anywhere? The process can't bind to port 80 without root privileges, which is unlikely given this is a test suite.

gmaclennan replied:
I'm using fastify.inject, which mocks this to `localhost:80`. On reflection I don't like this though: it requires reading the fastify docs and it's not a test of a real-world example. I'm going to change these tests to actually start the server on a port and do a proper request.

hackergrrl commented on src/backend/test/upgrade-server.test.js:
```diff
@@ -0,0 +1,204 @@
+  // @ts-ignore
+  expected1.url = `${baseUrl}/installers/${expected1.hash}`;
+
+  const apkFilepath = path.join(
+    __dirname,
+    "./fixtures/valid-apks/com.example.test_SDK21_VN1.0.0_VC1.apk"
+  );
```
Import `validApksFolder` ;)

hackergrrl commented on src/backend/test/upgrade-server.test.js:
```diff
@@ -0,0 +1,204 @@
+  await cleanup();
+
+  try {
+    await got(`${baseUrl}/installers`);
```
Test will still pass if this does not throw.

hackergrrl commented on src/backend/test/upgrade-server.test.js:
```diff
@@ -0,0 +1,204 @@
+  cleanup();
+
+  try {
+    await got(`${baseUrl}/installers`);
```
Test will still pass if this does not throw.

hackergrrl commented on src/backend/test/upgrade-server.test.js:
```diff
@@ -0,0 +1,204 @@
+
+  try {
+    // Server is stopped, this should throw
+    await got(`${baseUrl}/installers`);
```
Test will still pass if this does not throw.

hackergrrl reviewed (commented):
These are my final review comments. Great tests @gmaclennan! Very easy to read and follow. I like the way you've set it all up.

gmaclennan committed:
chore: Reduce noise in adb logcat logs

gmaclennan committed:
fix: Don't debug log in production

gmaclennan committed:
fix: Wait for server to start before trying to start p2p-upgrades

When starting the app directly to the sync screen the initial announce
was not happening

gmaclennan committed:
fix: Don't query self for upgrades

gmaclennan committed:
Don't use fastify.inject() for tests - test with actual requests via got

gmaclennan committed:
fix: Ensure storage tests fail if expected throw does not happen

gmaclennan committed:
chore: use readJson from helpers

gmaclennan committed:
feat: Show "No app updates found" once at least one device is checked

gmaclennan committed:
Add explanation for unusual code in upgrade-simulator

gmaclennan committed:
feat: Use keep-alive to re-use same connection for upgrade check polling

gmaclennan committed:
Add test for download completing before server close

gmaclennan committed:
Add test for broken connection

gmaclennan committed:
Cleanup AsyncService code and fix bugs

gmaclennan committed:
fix: Don't throttle state events, only download progress

The throttling seems to introduce race conditions in the tests (not all
events that the tests wait for are guaranteed to fire).

gmaclennan committed:
Fix manager tests

luandro committed:
Determine update candidate based on new version naming scheme (#596)

gmaclennan committed:
Remove debug logging on CI (was used for debugging test failure)

gmaclennan committed:
chore: Fix to backend build script

gmaclennan committed:
chore: Clean up logging code

gmaclennan committed:
Failing test for checkedPeers

gmaclennan committed:
fix: checkedPeers should only update after potential upgrades r checked

gmaclennan committed:
chore: typings for tests

gmaclennan committed:
failing test for checkedPeers reseting after stop,start

gmaclennan committed:
fix: Checked Peers should reset after start and stop

gmaclennan committed:
fix tests, stricter testing for empty arrays in state

the lodash.isMatch function we are using will match an empty array
against any array, so to check the state is an empty array, we need to
use the compare function to strictly check the array length.

gmaclennan commented:
I identified 2 bugs in my own QA testing:

- A device would show "No app updates found / checked 1 device" for a second or two before an update started downloading. E.g. it was reporting a device was checked before it had actually checked it.
- The checked devices list would not be reset when navigating away from and returning to the sync screen.

These bugs are fixed now in this PR, and there are more comprehensive tests for the "checkedPeers" functionality.

gmaclennan committed:
fix: Don't use same keep-alive connection for downloads and polling

fixes #612

gmaclennan committed:
Update upgrades QA build numbers

gmaclennan committed:
fix typo

gmaclennan committed:
Merge branch 'develop' into p2p-update/typescript

* develop:
  chore: Update instructions on NDK version in contributing docs
  chore: Update current NDK version in docs

gmaclennan committed:
Merge branch 'develop' into p2p-update/typescript

gmaclennan committed:
Merge branch 'develop' into p2p-update/typescript

gmaclennan committed:
Fix "Go to Map" button style for empty observation list screen

gmaclennan committed:
fix e2e debug test config (for local testing)

gmaclennan committed:
Increase server start timeout in attempt to fix e2e tests

gmaclennan committed:
Merge branch 'develop' into p2p-update/typescript
